From: Robert Marko <robert.marko@sartura.hr>
Date: Thu, 16 Jul 2026 12:00:00 +0200
Subject: [PATCH] board: airoha: add Datamate NanoDPU

Datamate NanoDPU currently shares the AN7583 evaluation board setup,
except that GDM1 is disabled and GDM2 uses SGMII without in-band status.
Enable GPT partition table support as well.

NanoDPU boots exclusively from eMMC, so omit the unused SPI flash, MTD
and UBI support.

Signed-off-by: Robert Marko <robert.marko@sartura.hr>
---
 arch/arm/dts/an7583-nanoDPU.dts  | 18 ++++++++
 configs/an7583_nanoDPU_defconfig | 72 ++++++++++++++++++++++++++++++++++++++++
 defenvs/an7583_nanoDPU_env       | 14 ++++++
 3 files changed, 104 insertions(+)
 create mode 100644 arch/arm/dts/an7583-nanoDPU.dts
 create mode 100644 configs/an7583_nanoDPU_defconfig
 create mode 100644 defenvs/an7583_nanoDPU_env

--- /dev/null
+++ b/arch/arm/dts/an7583-nanoDPU.dts
@@ -0,0 +1,18 @@
+// SPDX-License-Identifier: (GPL-2.0-only OR BSD-2-Clause)
+
+#include "an7583-evb.dts"
+
+/ {
+	model = "Datamate NanoDPU";
+	compatible = "datamate,nanoDPU", "airoha,an7583", "airoha,en7583";
+};
+
+&gdm1 {
+	status = "disabled";
+};
+
+&gdm2 {
+	phy-mode = "sgmii";
+	/delete-property/ managed;
+};
+
--- /dev/null
+++ b/configs/an7583_nanoDPU_defconfig
@@ -0,0 +1,72 @@
+CONFIG_ARM=y
+CONFIG_POSITION_INDEPENDENT=y
+CONFIG_ARCH_AIROHA=y
+CONFIG_TARGET_AN7583=y
+CONFIG_TEXT_BASE=0x81E00000
+CONFIG_SYS_MALLOC_F_LEN=0x4000
+CONFIG_NR_DRAM_BANKS=1
+CONFIG_ENV_SIZE=0x10000
+CONFIG_ENV_OFFSET=0x7c000
+CONFIG_DM_GPIO=y
+CONFIG_DEFAULT_DEVICE_TREE="an7583-nanoDPU"
+CONFIG_SYS_LOAD_ADDR=0x81800000
+CONFIG_BUILD_TARGET="u-boot.bin"
+# CONFIG_EFI_LOADER is not set
+CONFIG_FIT=y
+CONFIG_FIT_VERBOSE=y
+CONFIG_BOOTDELAY=3
+CONFIG_DEFAULT_FDT_FILE="an7583-nanoDPU"
+CONFIG_SYS_PBSIZE=1049
+CONFIG_SYS_CONSOLE_IS_IN_ENV=y
+# CONFIG_DISPLAY_BOARDINFO is not set
+CONFIG_HUSH_PARSER=y
+CONFIG_SYS_PROMPT="U-Boot> "
+CONFIG_SYS_MAXARGS=8
+CONFIG_CMD_BOOTZ=y
+CONFIG_CMD_BOOTMENU=y
+# CONFIG_CMD_ELF is not set
+# CONFIG_CMD_XIMG is not set
+CONFIG_CMD_BIND=y
+CONFIG_CMD_GPIO=y
+CONFIG_CMD_MMC=y
+# CONFIG_CMD_SETEXPR is not set
+CONFIG_CMD_MII=y
+CONFIG_CMD_DHCP=y
+CONFIG_CMD_PING=y
+CONFIG_EFI_PARTITION=y
+CONFIG_CMD_PART=y
+CONFIG_CMD_EXT4=y
+CONFIG_CMD_FAT=y
+CONFIG_CMD_FS_GENERIC=y
+CONFIG_CMD_LOG=y
+CONFIG_ENV_OVERWRITE=y
+CONFIG_ENV_IS_IN_MMC=y
+# CONFIG_ENV_IS_IN_MTD is not set
+CONFIG_ENV_MMC_USE_SW_PARTITION=y
+CONFIG_ENV_MMC_SW_PARTITION="ubootenv"
+CONFIG_ENV_RELOC_GD_ENV_ADDR=y
+CONFIG_ENV_USE_DEFAULT_ENV_TEXT_FILE=y
+CONFIG_ENV_DEFAULT_ENV_TEXT_FILE="defenvs/an7583_nanoDPU_env"
+CONFIG_ENV_VARS_UBOOT_RUNTIME_CONFIG=y
+CONFIG_NET_RANDOM_ETHADDR=y
+CONFIG_SYS_RX_ETH_BUFFER=8
+CONFIG_SYSCON=y
+CONFIG_CLK=y
+CONFIG_DMA=y
+CONFIG_LED=y
+CONFIG_LED_GPIO=y
+CONFIG_MMC_HS200_SUPPORT=y
+CONFIG_MMC_MTK=y
+CONFIG_DM_MDIO=y
+CONFIG_MDIO_AIROHA=y
+CONFIG_PCS_AIROHA_AN7583=y
+CONFIG_AIROHA_ETH=y
+CONFIG_PHY=y
+CONFIG_PINCTRL=y
+CONFIG_PINCTRL_AIROHA_AN7583=y
+CONFIG_POWER_DOMAIN=y
+CONFIG_DM_REGULATOR=y
+CONFIG_DM_REGULATOR_FIXED=y
+CONFIG_RAM=y
+CONFIG_DM_SERIAL=y
+CONFIG_SYS_NS16550=y
+CONFIG_SHA512=y
diff --git a/defenvs/an7583_nanoDPU_env b/defenvs/an7583_nanoDPU_env
new file mode 100644
--- /dev/null
+++ b/defenvs/an7583_nanoDPU_env
@@ -0,0 +1,14 @@
+loadaddr=0x81800000
+ipaddr=192.168.1.1
+serverip=192.168.1.10
+bootargs=console=ttyS0,115200 earlycon root=/dev/fit0 rootwait
+bootcmd=run boot_production
+bootfile_upg=openwrt-airoha-an7583-datamate_nanoDPU-squashfs-sysupgrade.itb
+part_default=production
+boot_production=run emmc_read_production && bootm $loadaddr
+boot_tftp_production=tftpboot $loadaddr $bootfile_upg && iminfo $loadaddr && run emmc_write_production
+mmc_read_vol=mmc read $loadaddr $part_addr 0x100 && imszb $loadaddr image_size && test 0x$image_size -le 0x$part_size && mmc read $loadaddr $part_addr $image_size
+mmc_write_vol=imszb $loadaddr image_size && test 0x$image_size -le 0x$part_size && mmc erase $part_addr $image_size && mmc write $loadaddr $part_addr $image_size
+emmc_read_production=part start mmc 0 $part_default part_addr && part size mmc 0 $part_default part_size && run mmc_read_vol
+emmc_write_production=part start mmc 0 $part_default part_addr && part size mmc 0 $part_default part_size && run mmc_write_vol
+boot_tftp=tftpboot $loadaddr $bootfile_upg && bootm $loadaddr
